### PR TITLE
Add NewMetaDataIterator method

### PR DESCRIPTION
**Pull request by mrambacher**
Fixes a problem where the iterator for metadata was being treated as a non-user key when in fact it was a user key.  This led to a problem where the property keys could not be searched for correctly.

The main exposure of this problem was that the HashIndexReader could not get the "prefixes" property correctly, resulting in the failure of retrieval/creation of the BlockPrefixIndex.

Added BlockBasedTableTest.SeekMetaBlocks test to validate this condition.

Fixing this condition exposed two other tests (SeekWithPrefixLongerThanKey, MultiGetPrefixFilter) that passed incorrectly previously and now failed.  Updated those two tests to pass.  Not sure if the tests are functionally correct/still appropriate, but made them pass...

### CODE REVIEW

**Comment by mrambacher** [comment]
Fixes #8673

File: table/block_based/block.cc [hunk 1]
```diff
@@ -968,6 +968,13 @@ Block::Block(BlockContents&& contents, size_t read_amp_bytes_per_bit,
   }
 }
 
+DataBlockIter* Block::NewMetaDataIterator() {
+  DataBlockIter* ret_iter =
+      NewDataIterator(BytewiseComparator(), kDisableGlobalSequenceNumber);
+  ret_iter->raw_key_.SetIsUserKey(true);
```
[the remark this answers is not in the document]
Comment (mrambacher) [reply table/block_based/block.cc]: There are other differences between a DataBlockIter and a MetaBlockIter:
- MetaBlockIter always use BytewiseComparator; DatBlockIter uses one associated with the table;
- MetaBlockIter uses non-assert functions for parsing keys (CheckAndDecodeEntry) whereas DataBlockIter asserts (DecodeEntry)
- MetaBlockIter will not change read amplification and is not reflected in the Statistics

The naming of "MetaBlockIter" was chosen as it is used to read from a variable called "metablock_index" in a method called "FindMetaBlock".  

If you prefer a different name, feel free to suggest one.

File: table/block_based/block.cc [hunk 2]
```diff
@@ -607,12 +675,12 @@ bool IndexBlockIter::ParseNextIndexKey() {
 // shared_size is 0, which included the first entry in each restart point.
 // Otherwise the format is delta-size = block handle size - size of last block
 // handle.
-void IndexBlockIter::DecodeCurrentValue(uint32_t shared) {
+void IndexBlockIter::DecodeCurrentValue(bool is_shared) {
   Slice v(value_.data(), data_ + restarts_ - value_.data());
   // Delta encoding is used if `shared` != 0.
```
Comment (pdillinger) [inline table/block_based/block.cc:680]: Update comment

**Review by pdillinger — COMMENTED** [review]
This needs performance testing. Changing templates (static polymorphism) to virtual functions (dynamic polymorphism) is probably a regression in this hot code, possibly also heap allocating objects that can go on the stack.

(Doesn't have to be here and now...) I think we're more likely to see net improvement if we change from an "iterator" interface to a "visitor" interface for meta blocks. Metaindex and properties should be parsed once on read, never on the fly. I don't know how much code that would involve, but naively it would seem less entangled with data and index block iteration.

**Comment by mrambacher** [comment]
> This needs performance testing. Changing templates (static polymorphism) to virtual functions (dynamic polymorphism) is probably a regression in this hot code, possibly also heap allocating objects that can go on the stack.
> 
> (Doesn't have to be here and now...) I think we're more likely to see net improvement if we change from an "iterator" interface to a "visitor" interface for meta blocks. Metaindex and properties should be parsed once on read, never on the fly. I don't know how much code that would involve, but naively it would seem less entangled with data and index block iteration.

@pdillinger I will run some performance tests but the Iterator class already has virtual functions.  This PR is not introducing any polymorphism.

File: table/block_based/block.cc [hunk 3]
```diff
@@ -709,9 +776,9 @@ bool BlockIter<TValue>::BinarySeek(const Slice& target, uint32_t* index,
     // The `mid` is computed by rounding up so it lands in (`left`, `right`].
     int64_t mid = left + (right - left + 1) / 2;
     uint32_t region_offset = GetRestartPoint(static_cast<uint32_t>(mid));
-    uint32_t shared, non_shared;
-    const char* key_ptr = DecodeKeyFunc()(
-        data_ + region_offset, data_ + restarts_, &shared, &non_shared);
+    uint32_t shared, non_shared, dummy;
+    const char* key_ptr = DecodeKV(data_ + region_offset, data_ + restarts_,
```
Comment (pdillinger) [inline table/block_based/block.cc:780]: DecodeKeyFunc -> DecodeKV (static -> dynamic polymorphism)

**Comment by mrambacher** [comment]
> This needs performance testing. Changing templates (static polymorphism) to virtual functions (dynamic polymorphism) is probably a regression in this hot code, possibly also heap allocating objects that can go on the stack.
> 
> (Doesn't have to be here and now...) I think we're more likely to see net improvement if we change from an "iterator" interface to a "visitor" interface for meta blocks. Metaindex and properties should be parsed once on read, never on the fly. I don't know how much code that would involve, but naively it would seem less entangled with data and index block iteration.

I ran the following command on both main and on this PR:
TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=filluniquerandom,compact -target_file_size_base=1048576 -compression_type=none
TEST_TMPDIR=/dev/shm ./db_bench -readonly=1 -use_existing_db=1 -benchmarks=seekrandom -seek_nexts=1000 -target_file_size_base=1048576 -cache_size=0 -duration=30

Results for main:
seekrandom   :     164.420 micros/op 6081 ops/sec;  672.5 MB/s (182999 of 182999 found)
seekrandom   :     165.281 micros/op 6050 ops/sec;  669.0 MB/s (181999 of 181999 found)
seekrandom   :     168.022 micros/op 5951 ops/sec;  658.1 MB/s (178999 of 178999 found)

Results for this PR:
seekrandom   :     165.602 micros/op 6038 ops/sec;  667.7 MB/s (181999 of 181999 found)
seekrandom   :     165.443 micros/op 6044 ops/sec;  668.3 MB/s (181999 of 181999 found)
seekrandom   :     169.142 micros/op 5912 ops/sec;  653.7 MB/s (177999 of 177999 found)

**Review by pdillinger — APPROVED** [review]
LGTM. I ran my own perf test as in #9126 and if anything, this is slightly faster than baseline, 122474 vs. 121924 ops/sec.

**Comment by pdillinger** [comment]
Note: FB test coverage indicates MetaBlockIter::SeekForPrevImpl is not covered.

**Comment by mrambacher** [comment]
> Note: FB test coverage indicates MetaBlockIter::SeekForPrevImpl is not covered.

I am not surprised this is not covered.  The method is required by the signature but everything that uses MetaBlocks either does a seek or scan.  I can delete the contents if you wish and have it assert/fail.

**Comment by pdillinger** [comment]
> I am not surprised this is not covered. The method is required by the signature but everything that uses MetaBlocks either does a seek or scan. I can delete the contents if you wish and have it assert/fail.

I don't have a specific preference

File: table/block_based/block.h [hunk 4]
```diff
@@ -582,12 +588,30 @@ class DataBlockIter final : public BlockIter<Slice> {
...
+    return value_;
+  }
+
+ protected:
+  virtual void SeekToFirstImpl() override;
+  virtual void SeekToLastImpl() override;
+  virtual void SeekImpl(const Slice& target) override;
+  virtual void SeekForPrevImpl(const Slice& target) override;
+  virtual void NextImpl() override;
+  virtual void PrevImpl() override;
```
Comment (riversand963) [inline table/block_based/block.h:614]: https://stackoverflow.com/questions/39932391/should-i-use-virtual-override-or-both-keywords
Please remove 'virtual'.

File: table/block_based/block.cc [hunk 5]
```diff
@@ -309,23 +359,21 @@ bool DataBlockIter::SeekForGetImpl(const Slice& target) {
...
-  } else {
-    limit = data_ + restarts_;
+  uint32_t limit = restarts_;
+  if (restart_index + 1 < num_restarts_) {
+    limit = GetRestartPoint(restart_index + 1);
   }
-
-  while (true) {
+  while (current_ < limit) {
+    bool shared;
```
Comment (riversand963) [inline table/block_based/block.cc:369]: Init

File: table/block_based/block.h [hunk 6]
```diff
@@ -582,12 +588,30 @@ class DataBlockIter final : public BlockIter<Slice> {
...
 
-  template <typename DecodeEntryFunc>
-  inline bool ParseNextDataKey(const char* limit = nullptr);
-
   bool SeekForGetImpl(const Slice& target);
-  void NextOrReportImpl();
-  void SeekToFirstOrReportImpl();
+};
+
+class MetaBlockIter final : public BlockIter<Slice> {
```
Comment (riversand963) [inline table/block_based/block.h:594]: Please add some comment describing why we need this class.

File: db/db_with_timestamp_basic_test.cc [hunk 7]
```diff
@@ -772,11 +772,13 @@ TEST_P(DBBasicTestWithTimestampTableOptions, SeekWithPrefixLessThanKey) {
...
 }
 
-TEST_P(DBBasicTestWithTimestampTableOptions, SeekWithPrefixLongerThanKey) {
+TEST_P(DBBasicTestWithTimestampTableOptions, SeekWithCappedPrefix) {
   Options options = CurrentOptions();
   options.env = env_;
   options.create_if_missing = true;
-  options.prefix_extractor.reset(NewFixedPrefixTransform(20));
+  // All of the keys or this test must be longer than 3 characters
+  static int kMinKeyLen = 3;
```
Comment (riversand963) [inline db/db_with_timestamp_basic_test.cc:780]: Nit: can be constexpr

File: table/block_based/block.cc [hunk 8]
```diff
@@ -462,7 +536,20 @@ void DataBlockIter::SeekToLastImpl() {
     return;
   }
   SeekToRestartPoint(num_restarts_ - 1);
-  while (ParseNextDataKey<DecodeEntry>() && NextEntryOffset() < restarts_) {
+  bool is_shared;
```
Comment (riversand963) [inline table/block_based/block.cc:539]: Nit: better to init

File: table/block_based/block.cc [hunk 9]
```diff
@@ -432,20 +480,46 @@ void DataBlockIter::SeekForPrevImpl(const Slice& target) {
...
 }
 
-void DataBlockIter::SeekToFirstOrReportImpl() {
+void MetaBlockIter::SeekToFirstImpl() {
   if (data_ == nullptr) {  // Not init yet
     return;
   }
   SeekToRestartPoint(0);
-  ParseNextDataKey<CheckAndDecodeEntry>();
+  bool is_shared;
```
Comment (riversand963) [inline table/block_based/block.cc:521]: better to init

File: table/block_based/block.cc [hunk 10]
```diff
@@ -432,20 +480,46 @@ void DataBlockIter::SeekForPrevImpl(const Slice& target) {
...
+  }
+}
+
 void DataBlockIter::SeekToFirstImpl() {
   if (data_ == nullptr) {  // Not init yet
     return;
   }
   SeekToRestartPoint(0);
-  ParseNextDataKey<DecodeEntry>();
+  bool is_shared;
```
Comment (riversand963) [inline table/block_based/block.cc:512]: better to init

File: table/block_based/block.cc [hunk 11]
```diff
@@ -153,6 +165,27 @@ void IndexBlockIter::PrevImpl() {
...
+    if (restart_index_ == 0) {
+      // No more entries
+      current_ = restarts_;
+      restart_index_ = num_restarts_;
+      return;
+    }
+    restart_index_--;
+  }
+  SeekToRestartPoint(restart_index_);
+  bool is_shared;
```
Comment (riversand963) [inline table/block_based/block.cc:182]: Nit: init

File: table/block_based/block.cc [hunk 12]
```diff
@@ -462,7 +536,20 @@ void DataBlockIter::SeekToLastImpl() {
...
+    // Keep skipping
+  }
+}
+
+void MetaBlockIter::SeekToLastImpl() {
+  if (data_ == nullptr) {  // Not init yet
+    return;
+  }
+  SeekToRestartPoint(num_restarts_ - 1);
+  bool is_shared;
```
Comment (riversand963) [inline table/block_based/block.cc:550]: Nit: init

File: table/block_based/block.cc [hunk 13]
```diff
@@ -126,10 +126,22 @@ struct DecodeKeyV4 {
   }
 };
 
-void DataBlockIter::NextImpl() { ParseNextDataKey<DecodeEntry>(); }
+struct DecodeEntryV4 {
+  inline const char* operator()(const char* p, const char* limit,
+                                uint32_t* shared, uint32_t* non_shared,
+                                uint32_t* value_length) {
+    *value_length = 0;
```
Comment (riversand963) [inline table/block_based/block.cc:133]: Nit: our internal linter may complain about nullptr ref. So assert(value_length).

File: table/block_based/block.h [hunk 14]
```diff
@@ -582,12 +588,30 @@ class DataBlockIter final : public BlockIter<Slice> {
...
+ public:
+  void Initialize(const char* data, uint32_t restarts, uint32_t num_restarts,
+                  bool block_contents_pinned) {
+    InitializeBase(BytewiseComparator(), data, restarts, num_restarts,
+                   kDisableGlobalSequenceNumber, block_contents_pinned);
+    raw_key_.SetIsUserKey(true);
```
Comment (riversand963) [inline table/block_based/block.h:600]: Given this is the core fix, maybe a line or two of explanation helps emphasizing the significance.

**Review by riversand963 — CHANGES_REQUESTED** [review]
I am not expert at block-related code, and rely on the author's benchmarking to demonstrate no regression on the hot path.

We follow https://google.github.io/styleguide/cppguide.html#Local_Variables, and as a good practice, we should always initialize variables. Relying on a subsequent function call to always set the value sounds risky.

**Review by riversand963 — APPROVED** [review]
LGTM. Thanks @mrambacher for the fix.

**Pull request closed without merging**